crates/bevy_asset/src/temp.rs

```diff
@@ -0,0 +1,56 @@
+use std::{
+    io::{Error, ErrorKind},
+    path::{Path, PathBuf},
+};
+
+use bevy_ecs::{system::Resource, world::World};
+
+use crate::io::AssetSourceBuilder;
+
+/// Private resource to store the temporary directory used by `temp://`.
+/// Kept private as it should only be removed on application exit.
+#[derive(Resource)]
+enum TempDirectory {
```

Comment:
nit: It seems a shame that `TempDirectory` is private since with this change we'll be going through the overhead of creating the directory every boot. I don't think this being public presents an obvious foot-gun, and having a temp directory around could be beneficial.

I guess the alternative mitigation is trying to defer temp directory creation until a temp asset is written, but that seems more technical work.

Reply:
So I experimented with trying to defer the creation of the `TempDir` (if required), but the way the `AssetSource` system is setup right now, combined with `TempDir` using RAII for automatic cleanup, makes deferring the creation fairly complex, and possibly more costly than just creating the folder up-front.

I _have_ made the `TempDirectory` resource public (but not in the prelude) to provide access to the `Path`, which may be helpful on desktop platforms. However, it will be largely useless on WASM so it might get removed anyway if/when a WASM compatible version of this PR is made.